src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -223,6 +213,114 @@ public static bool EncodeStringLiteral(string value, Span<byte> destination, out
+                    {
+                        throw new HttpRequestException(SR.net_http_request_invalid_char_encoding);
+                    }
+
+                    destination[i] = (byte)((uint)(c));
+                }
+
+                bytesWritten = value.Length;
+                return true;
+            }
+
+            bytesWritten = 0;
+            return false;
+        }
+
+        public static bool EncodeStringLiteral(string value, Span<byte> destination, out int bytesWritten)
+        {
+            // From https://tools.ietf.org/html/rfc7541#section-5.2
+            // ------------------------------------------------------
+            //   0   1   2   3   4   5   6   7
+            // +---+---+---+---+---+---+---+---+
+            // | H |    String Length (7+)     |
+            // +---+---------------------------+
+            // |  String Data (Length octets)  |
+            // +-------------------------------+
+
+            if (destination.Length != 0)
+            {
+                destination[0] = 0; // TODO: Use Huffman encoding
+                if (IntegerEncoder.Encode(value.Length, 7, destination, out int integerLength))
+                {
+                    Debug.Assert(integerLength >= 1);
+
+                    if (EncodeStringLiteralValue(value, destination.Slice(integerLength), out int valueLength))
+                    {
+                        bytesWritten = integerLength + valueLength;
+                        return true;
+                    }
+                }
+            }
+
+            bytesWritten = 0;
+            return false;
+        }
+
+        public static bool EncodeStringLiterals(string[] values, string separator, Span<byte> destination, out int bytesWritten)
+        {
+            bytesWritten = 0;
+
+            if (values.Length == 0)
+            {
+                return EncodeStringLiteral("", destination, out bytesWritten);
+            }
+            else if (values.Length == 1)
+            {
+                return EncodeStringLiteral(values[0], destination, out bytesWritten);
+            }
+
+            if (destination.Length != 0)
+            {
+                int valueLength = 0;
+
+                // Calculate length of all parts and separators.
+                foreach (string part in values)
+                {
+                    valueLength = checked((int)(valueLength + part.Length));
+                }
+
+                valueLength = checked((int)(valueLength + (values.Length - 1) * separator.Length));
+
+                if (IntegerEncoder.Encode(valueLength, 7, destination, out int integerLength))
+                {
+                    Debug.Assert(integerLength >= 1);
+
+                    int encodedLength = 0;
+                    for (int j = 0; j < values.Length; j++)
+                    {
+                        if (j != 0 && !EncodeStringLiteralValue(separator, destination.Slice(integerLength), out encodedLength))
```

Comment:
As a subsequent optimization, we could store the already-encoded separator bytes along with the header/separator information, and then here just blit it out rather than re-encoding it every time.